### PR TITLE
Enable running multiple glusterd2 instances per node

### LINKED ISSUE #186: Explore running multiple glusterd2 instances on a single node
If we could run multiple glusterd2 instances on a single node, this will simplify things like reproducer debuggability, writing and running functional tests etc.

TODOs to enable such functionality:

- [x] Make sure all of glusterd2 state can be stored in current working directory
- [x] Make all ports (etcd, rpc, rest) configurable via command line arguments
- [x] Allow add peer requests to specify ports in addition to IP address (like etcd does)
- [x] glusterd should return etcd peer url when it is being peer probed during peer add validation

### PR DESCRIPTION
**Pull request by prashanthpai**
Closes #186

### CODE REVIEW

**Comment by prashanthpai** [comment]
Bring up two glusterd2 instances, each in it's own working directory

```sh
[ppai@gd2-1 workdir]$ glusterd2 --restaddress 127.0.0.1:22007 --rpcaddress 127.0.0.1:22008 --etcdclientaddress 127.0.0.1:2279 --etcdpeeraddress 127.0.0.1:2280
```

```sh
[ppai@gd2-1 workdir2]$ glusterd2 --restaddress 127.0.0.1:23007 --rpcaddress 127.0.0.1:23008 --etcdclientaddress 127.0.0.1:2379 --etcdpeeraddress 127.0.0.1:2380
```

Each instance is now it's own standalone cluster

```sh

[ppai@gd2-1 ~]$ curl -s -X GET http://127.0.0.1:22007/v1/peers | python -m json.tool
[
    {
        "addresses": [
            "127.0.0.1:22008"
        ],
        "id": "fba61cee-9e8b-4815-b0bd-eef05995b5d6",
        "memberID": 12073872630648513194,
        "name": "gd2-1"
    }
]
```

```sh
[ppai@gd2-1 ~]$ curl -s -X GET http://127.0.0.1:23007/v1/peers | python -m json.tool
[
    {
        "addresses": [
            "127.0.0.1:23008"
        ],
        "id": "07c2a158-3379-4299-9d60-02ceac9640f3",
        "memberID": 13195394291058371180,
        "name": "gd2-1"
    }
]
```

Send a request to first instance to add second instance as peer.

```sh
[ppai@gd2-1 ~]$ cat addpeer.json 
{
	"addresses": ["127.0.0.1:23008"]
}
```

```sh
[ppai@gd2-1 ~]$ curl -X POST http://127.0.0.1:22007/v1/peers --data @addpeer.json -H 'Content-Type: application/json' 
{"id":"07c2a158-3379-4299-9d60-02ceac9640f3"}
```

List the peers from both instances and verify all is OK

```sh
[ppai@gd2-1 ~]$ curl -s -X GET http://127.0.0.1:22007/v1/peers | python -m json.tool
[
    {
        "addresses": [
            "127.0.0.1:23008"
        ],
        "id": "07c2a158-3379-4299-9d60-02ceac9640f3",
        "memberID": 9227756524249820644,
        "name": "gd2-1"
    },
    {
        "addresses": [
            "127.0.0.1:22008"
        ],
        "id": "fba61cee-9e8b-4815-b0bd-eef05995b5d6",
        "memberID": 12073872630648513194,
        "name": "gd2-1"
    }
]
```

```sh
[ppai@gd2-1 ~]$ curl -s -X GET http://127.0.0.1:23007/v1/peers | python -m json.tool
[
    {
        "addresses": [
            "127.0.0.1:23008"
        ],
        "id": "07c2a158-3379-4299-9d60-02ceac9640f3",
        "memberID": 9227756524249820644,
        "name": "gd2-1"
    },
    {
        "addresses": [
            "127.0.0.1:22008"
        ],
        "id": "fba61cee-9e8b-4815-b0bd-eef05995b5d6",
        "memberID": 12073872630648513194,
        "name": "gd2-1"
    }
]
```

Verify etcd cluster membership by querying both instances:

```sh
[ppai@gd2-1 ~]$ etcdctl --endpoint 127.0.0.1:2279 member list
800f93ab05b42de4: name=07c2a158-3379-4299-9d60-02ceac9640f3 peerURLs=http://127.0.0.1:2380 clientURLs=http://127.0.0.1:2379 isLeader=false
a78f032ffd1da2aa: name=fba61cee-9e8b-4815-b0bd-eef05995b5d6 peerURLs=http://127.0.0.1:2280 clientURLs=http://127.0.0.1:2279 isLeader=true
```

```sh
[ppai@gd2-1 ~]$ etcdctl --endpoint 127.0.0.1:2379 member list
800f93ab05b42de4: name=07c2a158-3379-4299-9d60-02ceac9640f3 peerURLs=http://127.0.0.1:2380 clientURLs=http://127.0.0.1:2379 isLeader=false
a78f032ffd1da2aa: name=fba61cee-9e8b-4815-b0bd-eef05995b5d6 peerURLs=http://127.0.0.1:2280 clientURLs=http://127.0.0.1:2279 isLeader=true
```

**TODO**: The `name` field for both peers is set to the hostname, hence the peer names are same. Test again with user specified peer names.

**Comment by prashanthpai** [comment]
Brick paths now can be of one of the following formats:

**Have both IP and port**
```json
	    "bricks": [
		"127.0.0.1:22008:/export/brick1/data",
		"127.0.0.1:22008:/export/brick2/data",
		"127.0.0.1:23008:/export/brick3/data",
		"127.0.0.1:22008:/export/brick4/data"
	    ],
```

**Just IP**
```json
	    "bricks": [
		"192.168.56.25:/export/brick1/data",
		"192.168.56.26:/export/brick2/data",
		"192.168.56.25:/export/brick3/data",
		"192.168.56.26:/export/brick4/data"
	    ],
```

**Just Peer UUID**
```json
	    "bricks": [
		"6ca433bb-6f51-4920-ad05-a59aec4acd4c:/export/brick1/data",
		"2132c520-4524-4ea0-a0dc-33884ad246c4:/export/brick2/data",
		"6ca433bb-6f51-4920-ad05-a59aec4acd4c:/export/brick3/data",
		"2132c520-4524-4ea0-a0dc-33884ad246c4:/export/brick4/data"
	    ],
```

File: commands/peers/addpeer.go [hunk 1]
```diff
@@ -31,27 +48,29 @@ func addPeerHandler(w http.ResponseWriter, r *http.Request) {
...
+	// and etcd peer url.
+	remotePeer, e := ValidateAddPeer(remotePeerAddress, &req)
 	if e != nil {
 		rest.SendHTTPError(w, http.StatusInternalServerError, remotePeer.OpError)
 		return
 	}
 
+	// TODO: Parse addresses considering ports to figure this out.
+	if isPeerInCluster(remotePeer.UUID) {
+		rest.SendHTTPError(w, http.StatusInternalServerError, errors.ErrPeerLocalNode.Error())
```
Comment (kshlm) [inline commands/peers/addpeer.go]: The error here should be that the new address is already a peer.

Also, is it possible to do a similar check before doing the RPC? GD1 does a similar thing.

For reference, GD1 does the following when a peer probe is done.

- Check is the given address is local. If it is return 'Probe not needed on localhost' error.
- Try to match the new address with known peer addresses.
  - If the exact same address exists, return error 'Peer already in cluster'
  - If peer match is found, just attach the new address to the matched peer, and send an update to the rest of the cluster. The matched peer isn't contacted at all.
- If a match isn't found, do a probe RPC (same as the `ValidateAddPeer` RPC here) to the address.
  - If the RPC returns an error, return the error to the CLI.
  - Check the returned peer id with existing peers
    - If a match exists, attach the new address to it
    - If not, a new peer is created.

This doesn't need to be replicated. We can simplify this or change this to make it better suitable for the new way of doing things.

File: commands/volumes/volume-create.go [hunk 2]
```diff
@@ -144,14 +144,32 @@ func nodesForVolCreate(req *volume.VolCreateRequest) ([]uuid.UUID, error) {
...
 		if err != nil {
 			return nil, err
 		}
-		id, err := peer.GetPeerIDByAddrF(addr)
-		if err != nil {
-			return nil, err
+
+		var id uuid.UUID
+		u := uuid.Parse(host)
+		if u != nil {
```
Comment (kshlm) [inline commands/volumes/volume-create.go]: You can directly assign to `id` and check it instead of using a temp var.

File: utils/peerutils.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+)
+
+const (
+	// DefaultRPCPort is the port on which glusterd2 instances will
+	// listen for incoming RPC requests when user hasn't explicitly
+	// set the port via config option.
+	DefaultRPCPort = 24008
+	// Accessing this as utils.DefaultRPCPort isn't neat. Just doing so
+	// to avoid cyclic imports for now. We may need a separate package
+	// other than 'gdctx' for global constants.
```
Comment (kshlm) [inline utils/peerutils.go]: Why not have it in config(viper)?

Comment (prashanthpai) [inline utils/peerutils.go]: I can do that too.

File: commands/peers/addpeer.go (hunk 1 above)
Comment (prashanthpai) [inline commands/peers/addpeer.go]: > The error here should be that the new address is already a peer.

 Will change that :+1:  Was an oversight when the local node check was modified.

> is it possible to do a similar check before doing the RPC

Yes. I've already captured this in the TODO comment above the call to `isPeerInCluster`. Will address that later

> Check is the given address is local. If it is return 'Probe not needed on localhost' error.

glusterd2 does this too. Right now, the error returned will be 'peer already in cluster' which is technically true.

> If peer match is found, just attach the new address to the matched peer, and send an update to the rest of the cluster. The matched peer isn't contacted at all.

I did not quite understand this.

> If a match isn't found, do a probe RPC (same as the ValidateAddPeer RPC here) to the address.
If the RPC returns an error, return the error to the CLI.

Already does this.

> Check the returned peer id with existing peers

Already does this as `isPeerInCluster` is called after `ValidateAddPeer`

> If a match exists, attach the new address to it
> If not, a new peer is created.

Multiple address per peer isn't handled here or elsewhere in the code. Requires a lot more effort.

**Pull request merged**